core/modules/replica/RebalanceJob.cc

```diff
@@ -649,39 +583,109 @@ void RebalanceJob::onJobFinish(MoveReplicaJob::Ptr const& job) {
         }
     }
 
-    // Evaluate the status of on-going operations to see if the job
-    // has finished.
+    // Try to submit one more job
 
-    size_t numLaunched, numFinished, numSuccess;
-    ::countJobStates(numLaunched, numFinished, numSuccess, _moveReplicaJobs);
+    size_t const numJobsLaunched = launchNextJobs(lock, 1);
+    if (numJobsLaunched != 0) {
+        _numLaunched += numJobsLaunched;
+    } else {
 
-    if (numFinished == numLaunched) {
-        if (numSuccess == numLaunched) {
+        // Evaluate the status of on-going operations to see if the job
+        // has finished.
 
-            // Make another iteration (and another one, etc. as many as needed)
-            // before it succeeds or fails.
-            //
-            // NOTE: a condition for this jobs to succeed is evaluated in
-            //       the precursor job completion code.
+        if (_numFinished == _numLaunched) {
+            finish(lock, _numSuccess == _numLaunched ? ExtendedState::SUCCESS
+                                                     : ExtendedState::FAILED);
+        }
+    }
+}
+
+size_t RebalanceJob::launchNextJobs(util::Lock const& lock,
+                                    size_t numJobs) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "launchNextJobs  numJobs=" << numJobs);
+
+    /*
+    size_t const numThreads =
+        controller()->serviceProvider()->config()->workerNumProcessingThreads();
+    */
+
+    // Compute the number of jobs which are already active at both ends
+    // (destination and source workers). Note that we need to scan both
+    // list to get a complete picture of what's on the workers' side.
 
-            restart(lock);
+    std::map<std::string,size_t> numAtDest;
+    std::map<std::string,size_t> numAtSrc;
 
+    for (auto&& ptr: _jobs) {
+        if (numAtDest.end() == numAtDest.find(ptr->destinationWorker())) {
+            numAtDest[ptr->destinationWorker()] = 0;
+        }
+        if (numAtSrc.end() == numAtSrc.find(ptr->sourceWorker())) {
+            numAtSrc[ptr->sourceWorker()] = 0;
+        }
+    }
+    for (auto&& ptr: _activeJobs) {
+        auto itrAtDest = numAtDest.find(ptr->destinationWorker());
+        if (numAtDest.end() == itrAtDest) {
+            numAtDest[ptr->destinationWorker()] = 1;
         } else {
-            finish(lock, ExtendedState::FAILED);
+            itrAtDest->second++;
+        }
```

Comment:
`numAtDest[ptr->destinationWorker()] ++`

Reply:
Why? This is a non-const iterator.

Reply:
I'm saying that this whole thing:

        auto itrAtDest = numAtDest.find(ptr->destinationWorker());
        if (numAtDest.end() == itrAtDest) {
            numAtDest[ptr->destinationWorker()] = 1;
        } else {
            itrAtDest->second++;
        }

is equivalent to `numAtDest[ptr->destinationWorker()] ++`

Reply:
Does the C++11 (or C++14) "zero initialization" rule applies to these counters? I'm reading through the documentation, and I'm not sure that's the case here:
https://en.cppreference.com/w/cpp/language/zero_initialization

Reply:
https://en.cppreference.com/w/cpp/container/map/operator_at
It is value-initialized (`T()`), same as zero-initialized for int type.

Reply:
👍